text/0000-pnpm-support.md

```diff
@@ -0,0 +1,270 @@
+
+Add first-class support for the [`pnpm`](https://pnpm.io) package manager.
+
+This will support everything we currently support for `npm` and `yarn`: pinning in `package.json`, auto-download, seamless version switching, global install interception, etc.
+This should be a feature-add, not a breaking change (we will not change `package.json` in an incompatible way).
+
+
+# Motivation
+[motivation]: #motivation
+
+pnpm is a popular Node package manager. At the time of this RFC, it has [11.9k stars on github](https://github.com/pnpm/pnpm), compared to [4.7k for npm](https://github.com/npm/cli) and [39.9k for yarn](https://github.com/yarnpkg/yarn).
+
+There are outstanding issues on the [pnpm project](https://github.com/pnpm/pnpm/issues/3146) and in [Volta](https://github.com/volta-cli/volta/issues/737) requesting that Volta support pnpm, and at least a few people have inquired about pnpm support on our Discord channel.
+
+Given that it's popular, and users want support for it, adding support for pnpm should further increase adoption of Volta, and additionally provide an example of how to onboard any new package managers we wish to support in the future.
+
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+We already support `yarn` as a 3rd-party package manager, so things should work the same way for `pnpm` (we shouldn't require any new terminology or concepts):
+
+- global installs (capturing commands and reworking them to install to our location)
+- cache expiry
+- hooks
+- autodownload
+- fetch / install / pin
+
+Using `volta run` should work the same way, with the addition of `--pnpm` and `--no-pnpm` options.
+
+## Documentation and Examples
+
+We will need to add examples and documentation on the website - this feature is not done until that happens.
+
+
+# Details
+[details]: #details
+
+## Fetch, Install, Pin
+
+We will need to duplicate the existing Yarn functionality (in `crates/volta-core/src/tool/yarn/`) for pnpm (into `crates/volta-core/src/tool/pnpm/`). This will likely be copy-paste with some modifications, and may be a chance for us to refactor some of this duplication that also exists for Npm.
+
+## Settings
+
+We will need to add support for reading and writing `pnpm` settings, mainly in the `platform`, `project`, and `toolchain` modules in `crates/volta-core/src/`. This should mostly involve adding the `pnpm` key alongside `npm` and `yarn`.
+
+This will modify the `package.json` format to add an additional optional `pnpm` field:
+
+```
+{
+  "volta": {
+    "node": 1.2.3,
+    "npm": 2.3.4,
+    "yarn": 3.4.5,
+    "pnpm": 4.5.6
+  }
+}
+```
+
+And we will need to add `--pnpm` and `--no-pnpm` to the options for the `volta run` command (in `src/command/run`).
+
+## Shims
+
+pnpm includes two executables, `pnpm` and `pnpx`. These will need to be supported in `crates/volta-core/src/run/`, the same way that we support `yarn` and `npx`. There is probably some opportunity for refactoring duplicated code in this section as well.
```

Comment:
We are deprecating pnpx in v7. I don't think support for pnpx is needed.

pnpx is changed to `pnpm dlx`

Reply:
Good to know, thanks!